common/types.go

```diff
@@ -234,16 +237,302 @@ func init() {
 }
 
 // Address represents the 20 byte address of an Ethereum account.
-type Address [AddressLength]byte
+type InternalAddress [AddressLength]byte
+type ExternalAddress [AddressLength]byte
+
+type Address struct {
+	inner AddressData
+}
+
+type AddressData interface {
+	Bytes() []byte
+	Hash() Hash
+	Hex() string
+	String() string
+	checksumHex() []byte
+	Format(s fmt.State, c rune)
+	SetBytes(b []byte)
+	MarshalText() ([]byte, error)
+	UnmarshalText(input []byte) error
+	UnmarshalJSON(input []byte) error
+	Scan(src interface{}) error
+	Value() (driver.Value, error)
+	Location() *Location
+	IsInChainScope() bool
+}
+
+func NewAddress() Address {
+	var a Address
+	var i InternalAddress
+	a.inner = &i
+	return a
+}
+
+// IsInChainScope checks if an address is a valid account in our node's sharded address space
+func (a Address) IsInChainScope() bool {
+	if a.inner == nil || a.inner.Hash() == ZeroAddr.Hash() {
+		return true
+	}
+	return NodeLocation.ContainsAddress(a)
+}
+
+func (a *Address) SetZero() *InternalAddress {
+	inner := &InternalAddress{0x0000000000000000000000000000000000000000}
+	a.inner = inner
+	return inner
+}
+
+func (a Address) InternalAddress() (*InternalAddress, error) {
+	if a.inner == nil {
+		return &InternalAddress{}, nil
+	}
+	internal, ok := a.inner.(*InternalAddress)
+	if !ok {
+		return nil, ErrInvalidScope
+	}
+	return internal, nil
+}
+
+func (a Address) Equals(b Address) bool {
+	if a.inner == nil && b.inner == nil {
+		return true
+	} else if a.inner == nil || b.inner == nil {
+		return false
+	}
+	return a.Hash() == b.Hash()
+}
 
 // BytesToAddress returns Address with value b.
 // If b is larger than len(h), b will be cropped from the left.
 func BytesToAddress(b []byte) Address {
+	if IsInChainScope(b) {
+		var i InternalAddress
+		i.SetBytes(b)
+		return Address{&i}
+	} else {
+		var e ExternalAddress
+		e.SetBytes(b)
+		return Address{&e}
+	}
+}
+
+func Bytes20ToAddress(b [20]byte) Address {
+	return BytesToAddress(b[:])
+}
+
+func InternalToAddress(i *InternalAddress) Address {
```

Comment:
(nit) Rather than having two functions `InternalToAddress()` and `ExternalToAddress()`, it would be more idiomatic to have one address constructor for this:
```
NewAddressFromData(inner AddressData) {return Address{inner: inner}}
```

Reply:
Fixed